src/site_apportionment/site_apportionment.py

```diff
@@ -0,0 +1,219 @@
+    df = df.merge(
+        dfb,
+        on=[ref, period],
+        how="left")
+    return df
+
+
+def weights(df):
+    """Calculates site weights based on the percents. Copies the precent value
+    from its original location (question 602) to a new column "site_percent".
+    If the percent value is Null, fills it with zero.
+
+    Adds all percents per RU reference and period and calculates
+    site_percent_total. If the total for a reference is zero, apportionment
+    cannot be done, and this reference is removed.
+    Re-noramalises the percent values by total, to compute site_weight. This
+    deals with the case when the users entered percents incorrectly, so they
+    don't sum up to 100.
+
+    Args:
+        df (pd.DataFrame): A dataframe containing all data
+
+    Returns:
+        (pd.DataFrame): A copy of original dataframe with an additional column
+        called site_weight countaining the weights of each site, between 0 and 1
+    """
+    dfc = df.copy()
+    dfc["site_percent"] = dfc[percent]
+    dfc["site_percent"].fillna(0, inplace=True)
+
+    # Set the percent for instance 0 to be 0
+    dfc["site_percent"] = dfc["site_percent"] * dfc[ins].astype("bool")
+
+    # Calculate the total percent for each reference and period
+    dfc["site_percent_total"] = (
+        dfc.groupby([ref, period])["site_percent"].transform("sum"))
+
+    # Filter out the rows where total percent is zero
+    dfc = dfc[dfc["site_percent_total"] != 0]
+
+    # Compute weights
+    dfc["site_weight"] = dfc["site_percent"] / dfc["site_percent_total"]
+
+    # Remove unnecessary
+    dfc.drop(columns=['site_percent', 'site_percent_total'], inplace=True)
+
+    return dfc
+
+
+def apportion_sites(df: pd.DataFrame)-> pd.DataFrame:
+    """Apportions the numerical values for each product group across multiple
+    sites, using percents from question 602 to compute weights.
+    Selects the records long from references that have multiple non-empty
+    postcodes. Splits the dataframe in two, codes and sites. Codes have
+    reference, period, product group (unique combinations of product, civil or
+    defence and pg_numeric) and (all numeric columns. Sites have reference,
+    period and all other fields except for product group keys and numerical
+    values. Sites dataframe contains multiple instances, with instance 1 and
+    higher having different postcodes and with percents for each site.
+    For sites, weights are calculated using the percents. Then, a Cartesian
+    product of product groups and sites is created, and the weights of each site
+    are applied to values of each product. Also, for short forms, sets percent
+    to 100.
+
+    Args:
+        df (pd.DataFrame): Dataframe containing all input data
+
+    Returns:
+        (pd.DataFrame): A dataframe with the same columns, with applied site
+        apportionment
+    """
+
+    # Value columns that we want to apportion
+    want_cals = [str(x) for x in range(202, 509)]
```

Comment:
Do you mean "cols" in `want_cals`? 

Also these numbers are hard coded. Would be better to provide them as an argument. Maybe as a list from a config file?

Reply:
We got rid of want_cols. The columns that we want are now coming from the config.